File: src/commonMain/kotlin/com/pricecompare/pricing/DiscountCalculator.kt
package com.pricecompare.pricing

import com.pricecompare.model.Decimal
import com.pricecompare.model.Discount
import com.pricecompare.model.Money
import com.pricecompare.util.Rounding

/**
 * src/commonMain/kotlin/com/pricecompare/pricing/DiscountCalculator.kt
 *
 * 割引適用を担当するユーティリティ。
 * 割引リストを適用順序（リスト順）に処理し、合計割引額を返す。
 *
 * 関連ファイル:
 * - src/commonMain/kotlin/com/pricecompare/model/Discount.kt
 * - src/commonMain/kotlin/com/pricecompare/model/Money.kt
 * - src/commonMain/kotlin/com/pricecompare/model/Offer.kt
 */
object DiscountCalculator {

    /**
     * 商品割引（Percentage, FixedAmount）のみを適用する。
     * クーポン、ポイントは別途処理。
     */
    fun applyItemDiscounts(baseAmount: Money, discounts: List<Discount>): DiscountResult {
        var currentAmount = baseAmount
        var totalDiscount = Money.ZERO

        for (discount in discounts) {
            when (discount) {
                is Discount.PercentageDiscount,
                is Discount.FixedAmount -> {
                    val discountAmount = discount.calculateAmount(currentAmount)
                    currentAmount = currentAmount - discountAmount
                    totalDiscount = totalDiscount + discountAmount
                }
                // クーポンとポイントは item discounts では無視
                is Discount.CouponDiscount,
                is Discount.PointRedemption -> {}
            }
        }

        return DiscountResult(
            discountedAmount = Rounding.roundMoney(currentAmount.amount),
            totalDiscount = Rounding.roundMoney(totalDiscount.amount)
        )
    }

    /**
     * クーポン割引を適用する。
     */
    fun applyCoupons(baseAmount: Money, coupons: List<Discount.CouponDiscount>): DiscountResult {
        var currentAmount = baseAmount
        var totalCoupon = Money.ZERO

        for (coupon in coupons) {
            val discountAmount = coupon.calculateAmount(currentAmount)
            currentAmount = currentAmount - discountAmount
            totalCoupon = totalCoupon + discountAmount
        }

        return DiscountResult(
            discountedAmount = Rounding.roundMoney(currentAmount.amount),
            totalDiscount = Rounding.roundMoney(totalCoupon.amount)
        )
    }

    /**
     * 割引適用結果を保持するデータクラス。
     */
    data class DiscountResult(
        val discountedAmount: Money,
        val totalDiscount: Money
    )
}

File: flutter_app/lib/engine/discount_calculator.dart
import 'money.dart';
import 'offer.dart';
import 'rounding.dart';

class DiscountResult {
  final Money discountedAmount;
  final Money totalDiscount;

  const DiscountResult({
    required this.discountedAmount,
    required this.totalDiscount,
  });
}

abstract final class DiscountCalculator {
  static DiscountResult applyItemDiscounts(
    Money baseAmount,
    List<Discount> discounts,
  ) {
    var currentAmount = baseAmount;
    var totalDiscount = Money.zero;

    for (final discount in discounts) {
      if (discount is FixedAmountDiscount ||
          discount is PercentageDiscount) {
        final discountAmount = discount.calculateAmount(currentAmount);
        currentAmount = currentAmount - discountAmount;
        totalDiscount = totalDiscount + discountAmount;
      }
    }

    return DiscountResult(
      discountedAmount: Rounding.roundMoney(currentAmount.amount),
      totalDiscount: Rounding.roundMoney(totalDiscount.amount),
    );
  }

  static DiscountResult applyCoupons(
    Money baseAmount,
    List<CouponDiscount> coupons,
  ) {
    var currentAmount = baseAmount;
    var totalCoupon = Money.zero;

    for (final coupon in coupons) {
      final discountAmount = coupon.calculateAmount(currentAmount);
      currentAmount = currentAmount - discountAmount;
      totalCoupon = totalCoupon + discountAmount;
    }

    return DiscountResult(
      discountedAmount: Rounding.roundMoney(currentAmount.amount),
      totalDiscount: Rounding.roundMoney(totalCoupon.amount),
    );
  }
}
